### DISCUSSION TITLE
Default mapping and gestures - Analysis and ideas

### DISCUSSION THREAD

**Discussion by ManDay** [8VIM/8VIM#249] [Ideas]
Please provide your feedback on the following idea for the **default** layout and gestures! The possibility that in the future one can freely map to both (layout is already possible) notwithstanding, of course.

For brevity, I shall abbreviate the sectors (`[:sector:]`) as `T` (`Top`), `R` (`Right`), `L` (`Left`), `B` (`Bottom`), and `C` (`Inside_circle`) in regex:

The typing sequences *from touch to release* are currently `(C[[:sector:]]+)+C([[:sector:]]*)` where the second capture group of trailing `[[:sector:]]`s, if not empty, only has the effect of omitting a trailing `<space>` but is not further differentiated. This could potentially be changed (for example, only one specific sector omits the trailing space), but I think it should be kept.

The above leaves all sequences which
 1. **don't start in the circle** `([[:sector:]]+C)*[[:sector:]]*`
  The 12 most simple of these are lines which start at a sector and terminate immediately (**tap the sector**), go to the circle (**swipe inwards**), or cross the circle into the opposing sector (**swipe across**). I propose the following bindings to them:
  - `T`: Shift (*unchanged*)
  - `TC`: Copy (mnemonic: pull the contents towards you, off the screen)
  - `TCB`: Cut (mnemonic: a "stronger" and longer version of *Copy*)
  - `R`: Backspace (*unchanged*)
  - `RC`: Select all (mnemonic: the cursor is at the end of the text and you want to select everything before)
  - `RCL`: Pos1 (mnemonic: all the way back)
  - `L`: Number-Pad or more generally the grid-layout (*unchanged*)
  - `LC`: Tab (mnemonic: a bit like `CR`)
  - `LCR`: End (mnemonic: opposite of *Pos1*; all the way forth)
  - `B`: Return (*unchanged*)
  - `BC`: Paste (mnemonic: push the clipboard onto the screen)
  - `BCT`: Emoji-Keyboard (mnemonic: paste an Emoji)
  
    NB on changes: `TCB` is currently meant to hide the keyboard, but this is not needed because Android provides a button for it, as far as I know. `RC` is currently bound to bringing up the selection keyboard, all of whose functions would now be available as gestures. The selection keyboard could still be opened with the sidebar button.
 2. **never return to the circle** `C[[:sector:]]+`
  The 4 most simple of these -- besides the tap in the circle for `<space>` -- are the ones which swipe outside. They are currently bound to the direction arrows (move the cursor), and I think this is a good choice which should be kept.

With these bindings, all the current sidebar buttons would be optional:
 - The Emoji-Keyboard is available as a gesture
 - Tab is available both, as a dedicated gesture and from the `CR` gesture if the cursor is at the end of an input field
 - All functions in the selection keyboard are available as gesture
 - Settings can be accessed through Android System Settings

For aid, the three gestures per sector from the 1st group could be indicated as three icons (instead of one icon currently).

**Comment by Gorkycreator** [comment]
Android does provide buttons to hide the keyboard, but they are not always visible (an on screen, keyboard-specific button does not show up with gesture controls), nor clear (shared functionality with the back button), so I've found the current use of TCB to be surprisingly useful. Keeping that functionality would be preferred, even if it's shifted to an odd gesture like RCB ("backspace the keyboard down").

The proposed use for RCL and LCR overlap a bit with CT and CB, as well as just tapping at the beginning and end of the text and correcting if it's off by a few characters with the keyboard. I'd propose placing Pos1 and End on gestures like Shift+CT or CTC. That would keep all cursor manipulation originating from C. This could also permit intuitive home and end key functionality with CRC/CLC.

I hadn't noticed this before playing around with functionality for this post, but selecting and launching into the selection keyboard from RC+hold is very useful. Maybe this could be preserved and expanded with RCL as back select and LCR as forward select gestures?

**Comment by ManDay** [comment]
Thank you for the feedback.

> I hadn't noticed this before playing around with functionality for this post, but selecting and launching into the selection keyboard from RC+hold is very useful.

I've wrongly written `LC` would currently bring up the selection keyboard. Of course, this is supposed to be `RC`, as you rightfully wrote there!

> Android does provide buttons to hide the keyboard, but they are not always visible [...]

I am not experienced with the variety of Androids and their UX/UI, so I appreciate that new perspective on this. In general, I'd say that if, "in theory", Android is supposed to provide that function and is just bad at doing so (for me it works well), then incorporating that function into the keyboard itself is a work-around. A work-around, which is hopefully only temporary in nature until Android improves to live up to "the theory". Hence, I think putting "hide keyboard" on a less optimal gesture such as `RCB` is a good idea.

> The proposed use for RCL and LCR overlap a bit with CT and CB, [...].  Maybe this could be preserved and expanded with RCL as back select and LCR as forward select gestures?

Although I do not see how said "overlap" is a problem (mind to elaborate?), I think you point out a good alternative. I would rather **not** bind to `C[[:sector:]]C`, though, because of #240, but binding to Shift-`CT` and Shift-`CB` seems like a viable alternative. `Pos1` and `End` are relatively rare and requiring a tap on Shift wouldn't be a worry. Meanwhile, being able to forward and backward select sounds terribly useful! I think Shift-`C[RL]` is a bit more intuitive because I associate `Pos1` and `End` with horizontal movements, but either of those is fine.

> but selecting and launching into the selection keyboard from RC+hold is very useful.

I can not relate to this, because either you are a "power user" and make full use of the gestures, in which case all the functions of the selection keyboard are already accessible. Or you are not comfortable with the gestures and you want buttons of the selection keyboard; but in that case you have the dedicated "Selection-Keyboard" button. Do you require a gesture to open the selection keyboard, if you're not using gestures to perform the functions therein?

**Comment by Gorkycreator** [comment]
> Although I do not see how said "overlap" is a problem (mind to elaborate?)

It's not really a problem, I was just considering more optimal use of RCL/LCR. Those are prime gestures, if you will, and might be more suited for more common use cases. (further opinions included below)

> ...either you are a "power user" and make full use of the gestures, in which case all the functions of the selection keyboard are already accessible. 

The gestures do not handle selection of text well, unfortunately, and I don't think there's an easy way to mimic the arrows from the selection keyboard--particularly the ability to toggle between tweaking  the beginning or end of the active selection. The current functionality of RC allows me to select the last sentence I wrote by holding the gesture and launches me directly into the selection keyboard upon release so I can tweak for precision (useful because timing the end of a selection is tricky).

I just thought this was a nifty feature and am interested in its preservation. I'll definitely be customizing my layout as soon as I find documentation, so I'm not overly concerned about these ideas making it into the defaults.

> Do you require a gesture to open the selection keyboard, if you're not using gestures to perform the functions therein?

If the goal of builiding out full gesture support is to pave the way for an option to hide the sidebar buttons, I would find it handy!

**Comment by ManDay** [comment]
> The gestures do not handle selection of text well, unfortunately, and I don't think there's an easy way to mimic the arrows from the selection keyboard--particularly the ability to toggle between tweaking the beginning or end of the active selection.

I see. What do you think about the following mappings, then?
 - `RCL`: Extend selection towards the left
 - `LCR`: Extend selection towards the right
 - `CL`: Shrink selection at the right end; if 0 wide (only a cursor), move the cursor left
 - `CR`: Shrink selection at the left end; if 0 wide (only a cusor), move the cusor right
 - Shift-`CL`: Place cursor at left end of selection; if 0 wide, act as `Pos1`
 - Shift-`CR`: Place cursor at right end of selection; if 0 wide, act as `End`

If that's not satisfactory, could you describe a sequence of inputs where you'd require more?

**Comment by Gorkycreator** [comment]
Those mappings sound good. If I understand correctly, the premise is RCL and LCR start a selection and the cursor movements switch function to help adjust the selection (by reducing the selection in your example). Shift+CL breaks the selection and leaves the cursor at the start point, and Shift+CR at the end point.

This covers most cases and would greatly reduce the need for the selection keyboard. The last use case the selection keyboard would perform better for is selecting larger blocks of text, like paragraphs (via the up and down arrows).

To achieve full parity, I'd propose shifting things around a bit to adapt LCR, RCL, TCB, and BCT as "cursor+" gestures. Here's the idea:

Without selection:
C* -> standard cursor movements, hold to repeat
LCR, RCL -> home, end (of current line); hold to start selection left or right
BCT, TCB -> pos1, end (of text); hold to start selection up or down

With selection:
C* -> move end selection position left, right, up, or down; hold to repeat
\*C\* -> move start selection left, right, up, or down; hold to repeat
Shift+C* -> break selection and jump to start[LT]/end[RB]

With anything: 
TLBRT -> emoji keyboard
Shift+TC -> cut

Once again, this is just throwing ideas out there. I feel there are a few compromises in this proposal, one being contextual copy/cut, which may cause unintended mistakes. The cursor+ gestures may also be jarring with a forward select functionality, initially capturing a selection from posA to posB and then "switching" to control the selection at posA.

I hope this is helpful feedback. I can definitely work with any of the proposed changes, and getting a copy function in the gestures will by itself be a huge QOL improvement!
